Index: openswan-2.4.0rc4.quilt/programs/pluto/demux.c
===================================================================
--- openswan-2.4.0rc4.quilt.orig/programs/pluto/demux.c	2006-12-03 19:27:28.000000000 +0200
+++ openswan-2.4.0rc4.quilt/programs/pluto/demux.c	2006-12-03 19:27:36.000000000 +0200
@@ -2585,6 +2585,9 @@
 
 		case EVENT_NULL:	/* non-event */
 		case EVENT_REINIT_SECRET:	/* Refresh cookie secret */
+#if 1 /* l2tp-workaround: do not panic on every EVENT_REINIT_SECRET. */
+		    break;
+#endif
 		default:
 		    bad_case(kind);
 		}
