Index: openswan-2.4.0rc4.quilt/programs/pluto/ipsec_doi.c
===================================================================
--- openswan-2.4.0rc4.quilt.orig/programs/pluto/ipsec_doi.c	2006-12-03 19:33:19.000000000 +0200
+++ openswan-2.4.0rc4.quilt/programs/pluto/ipsec_doi.c	2006-12-03 19:33:22.000000000 +0200
@@ -3760,7 +3760,7 @@
 
 #ifdef NAT_TRAVERSAL
     if (st->hidden_variables.st_nat_traversal) {
-      if (!out_vendorid(auth_payload
+      if (!out_vendorid(ISAKMP_NEXT_NONE
 			, &md->rbody
 			, md->quirks.nat_traversal_vid)) {
 	return STF_INTERNAL_ERROR;
@@ -3768,7 +3768,7 @@
     }
 
     if (st->hidden_variables.st_nat_traversal & NAT_T_WITH_NATD) {
-      if (!nat_traversal_add_natd(auth_payload, &md->rbody, md))
+      if (!nat_traversal_add_natd(ISAKMP_NEXT_NONE, &md->rbody, md))
 	return STF_INTERNAL_ERROR;
     }
 #endif
